jax_galsim/wcs.py

```diff
@@ -0,0 +1,922 @@
+
+
+class EuclideanWCS(BaseWCS):
+    """A EuclideanWCS is a `BaseWCS` whose world coordinates are on a Euclidean plane.
+    We usually use the notation (u,v) to refer to positions in world coordinates, and
+    they use the class `PositionD`.
+    """
+
+    # All EuclideanWCS classes must define origin and world_origin.
+    # Sometimes it is convenient to access x0,y0,u0,v0 directly.
+    @property
+    def x0(self):
+        """The x component of self.origin."""
+        return self.origin.x
+
+    @property
+    def y0(self):
+        """The y component of self.origin."""
+        return self.origin.y
+
+    @property
+    def u0(self):
+        """The x component of self.world_origin (aka u)."""
+        return self.world_origin.x
+
+    @property
+    def v0(self):
+        """The y component of self.world_origin (aka v)."""
+        return self.world_origin.y
+
+    def xyTouv(self, x, y, color=None):
+        """Convert x,y from image coordinates to world coordinates.
+
+        This is equivalent to ``wcs.toWorld(x,y)``.
+
+        It is also equivalent to ``wcs.posToWorld(galsim.PositionD(x,y))`` when x and y are scalars;
+        however, this routine allows x and y to be numpy arrays, in which case, the calculation
+        will be vectorized, which is often much faster than using the pos interface.
+
+        Parameters:
+            x:          The x value(s) in image coordinates
+            y:          The y value(s) in image coordinates
+            color:      For color-dependent WCS's, the color term to use. [default: None]
+
+        Returns:
+            ra, dec
+        """
+        if color is None:
+            color = self._color
+        return self._xyTouv(x, y, color=color)
+
+    def uvToxy(self, u, v, color=None):
+        """Convert u,v from world coordinates to image coordinates.
+
+        This is equivalent to ``wcs.toWorld(u,v)``.
+
+        It is also equivalent to ``wcs.posToImage(galsim.PositionD(u,v))`` when u and v are scalars;
+        however, this routine allows u and v to be numpy arrays, in which case, the calculation
+        will be vectorized, which is often much faster than using the pos interface.
+
+        Parameters:
+            u:          The u value(s) in world coordinates
+            v:          The v value(s) in world coordinates
+            color:      For color-dependent WCS's, the color term to use. [default: None]
+        """
+        if color is None:
+            color = self._color
+        return self._uvToxy(u, v, color)
+
+    # Simple.  Just call _u, _v.
+    def _posToWorld(self, image_pos, color):
+        x = image_pos.x - self.x0
+        y = image_pos.y - self.y0
+        return PositionD(self._u(x, y, color), self._v(x, y, color)) + self.world_origin
+
+    def _xyTouv(self, x, y, color):
+        x = x - self.x0  # Not -=, since don't want to modify the input arrays in place.
+        y = y - self.y0
+        u = self._u(x, y, color)
```

Comment:
where is `_u` defined??

Reply:
or is `EuclideanWCS` supposed to be an 'abstract' class?

Reply:
It's not defined in the original GalSim either so I'm going to ignore this for now @b-remy